app/src/main/java/com/poupa/vinylmusicplayer/misc/queue/StaticPlayingQueue.java

```diff
@@ -0,0 +1,458 @@
+        position = position + 1;
+
+        int n = songs.size() - 1;
+        for (int i = n; i >= 0; i--) {
+            originalQueue.add(previousPosition, new IndexedSong(songs.get(i), previousPosition + i));
+
+            queue.add(position, new IndexedSong(songs.get(i), previousPosition + i));
+
+            if (position < this.currentPosition) {
+                this.currentPosition++;
+            }
+        }
+
+        for (int i = 0; i < queue.size(); i++) {
+            originalQueue.get(i).index = i;
+            if (!(i >= position && i <= position+n) && queue.get(i).index >= previousPosition  ) {
+                queue.get(i).index = queue.get(i).index + n + 1;
+            }
+        }
+
+        songsIsStale = true;
+    }
+
+    // move song from from to to, position are conserved
+    public void move(int from, int to) {
+        if (from == to) return;
+        final int currentPosition = this.currentPosition;
+
+        IndexedSong songToMove = queue.remove(from);
+        queue.add(to, songToMove);
+
+        if (shuffleMode == SHUFFLE_MODE_NONE) {
+            IndexedSong previousSongToMove = originalQueue.remove(from);
+            originalQueue.add(to, previousSongToMove);
+
+            for (int i = 0; i < queue.size(); i++) {
+                queue.get(i).index = i;
+                originalQueue.get(i).index = i;
+            }
+        }
+
+        if (from > currentPosition && to <= currentPosition) {
+            this.currentPosition = currentPosition + 1;
+        } else if (from < currentPosition && to >= currentPosition) {
+            this.currentPosition = currentPosition - 1;
+        } else if (from == currentPosition) {
+            this.currentPosition = to;
+        }
+
+        songsIsStale = true;
+    }
+
+    private int rePosition(int deletedPosition) {
+        int position = this.currentPosition;
+
+        if (deletedPosition < position) {
+            this.currentPosition = position - 1;
+        } else if (deletedPosition == position) { //the current position was deleted
+            if (queue.size() > deletedPosition) {
+                return position;
+            } else {
+                return position - 1;
+            }
+        }
+
+        return INVALID_POSITION;
+    }
+
+    // remove song at index position, numbering need to be redone for every song after this position (-1)
+    public int remove(int position) {
+        IndexedSong o = queue.remove(position);
+        originalQueue.remove(o.index);
+
+        for (int i = 0; i < queue.size(); i++) {
+            originalQueue.get(i).index = i;
+            if (queue.get(i).index >= o.index) {
+                queue.get(i).index = queue.get(i).index - 1;
+            }
+        }
+
+        songsIsStale = true;
+
+        return rePosition(position);
+    }
+
+    private int removeAllOccurrence(Song song) {
+        int hasPositionChanged = INVALID_POSITION;
+
+        for (int i = queue.size() - 1; i >= 0; i--) {
```

Comment:
is there a particular reason why the queue is traversed from the end to the beginning?